### PR TITLE
New MISALIGNED LDST Test

### PR DESCRIPTION
**Pull request by neelgala**
The current I-MISALIGNED_LDST-01.S test in the rv32i test-suite generates a signature based on the trap behavior. However, this test assumes that all implementations will trap on mis-aligned data accesses. Implementations which support misaligned accesses in HW will never trap and thus the signatures will not match. 

This pull request includes a new MISLAIGNED_LDST test which will generate the same signature if an implementations support misaligned access or traps on such an access. It uses a simple handler which implements the required misaligned access. 

The signature is independent of whether a trap is taken or not.

ChangeLog:
2019-01-24 Neel Gala <neelgala@incoresemi.com>
    * riscv-test-suite/rv32i/src/I-MISALIGN_LDST-02.S: New test added to check 
    mis-aligned data accesses. This test uses only mtvec and mepc does not 
    validate the trap mechanism.
    * riscv-test-suite/rv32/references/I-MISALIGN_LDST-02.reference_output: 
    Added reference for above mentioned test.
    * riscv-test-suite/rv32i/Makefrag: replaced old MISALIGNED_LDST-01 test 
    with new test.
    * riscv-test-suite/rv32i/src/I-MISALIGN_LDST-01.S: removed old test src
    * riscv-test-suite/rv32i/references/I-MISALIGN_LDST-01.reference_output:
		removed old test reference.

related to #22

### CODE REVIEW

File: riscv-test-suite/rv32i/Makefrag [hunk 1]
```diff
@@ -33,7 +33,7 @@ rv32i_sc_tests =    \
     I-RF_size-01 \
     I-RF_width-01 \
     I-MISALIGN_JMP-01 \
-    I-MISALIGN_LDST-01 \
+    I-MISALIGN_LDST-02 \
```
[the remark this answers is not in the document]
Comment (neelgala) [reply riscv-test-suite/rv32i/Makefrag]: Where would you like me to move the original test? I think it is best suited for a "Machine" -trap check environment. Not sure where in the current directory structure would that test fit in

File: riscv-test-suite/rv32i/src/I-MISALIGN_LDST-02.S [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    # Test - mis-aligned by 1 byte
+    addi    x3,  x3,  0x1
+    li      x4,  0
+    lh      x4, 0(x3)
+    sw      x4, 4(x2)
+
+    # Test - should pass on all implementations 
+    addi    x3,  x3,  0x1
+    li      x4,  0
+    lh      x4, 0(x3)
```
[the remark this answers is not in the document]
Comment (neelgala) [reply riscv-test-suite/rv32i/src/I-MISALIGN_LDST-02.S]: The test is the same as that of the original test- I-MISALIGN_LDST-01. 
You are right - it is a sanity check which is not required here. Will remove it.

File: riscv-test-suite/rv32i/src/I-MISALIGN_LDST-02.S [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    sw      x4, 8(x2)
+
+    # Test - mis-aligned by 1 byte
```
[the remark this answers is not in the document]
Comment (neelgala) [reply riscv-test-suite/rv32i/src/I-MISALIGN_LDST-02.S]: Thanks for the catch. Will fix the comments in both LH and LHU.

File: riscv-test-suite/rv32i/src/I-MISALIGN_LDST-02.S [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    # restore mtvec and jump to the end
+    csrw mtvec, x31
+    jal x0, test_end
+
+    RVTEST_IO_WRITE_STR("# Test part A5  - Complete\n");
+
+    # ---------------------------------------------------------------------------------------------
+    # Exception handler
+lw_trap_handler:
+    # load fourth byte
```
[the remark this answers is not in the document]
Comment (neelgala) [reply riscv-test-suite/rv32i/src/I-MISALIGN_LDST-02.S]: yeah.. I would have to check if x4 is still 0 as soon as I enter the load-trap handlers.
Will submit a pull request with changes.
Thanks,

File: riscv-test-suite/rv32i/Makefrag (hunk 1 above)
[the remark this answers is not in the document]
Comment (hajekra) [reply riscv-test-suite/rv32i/Makefrag]: It's not about moving the test. If CTG decides to use complex test approach, the original test should be simply replaced by new implementation. Old test would not be used anyway and you will always have history if you want to see the old version.

File: riscv-test-suite/rv32i/src/I-MISALIGN_LDST-02.S (hunk 2 above)
[the remark this answers is not in the document]
Comment (hajekra) [reply riscv-test-suite/rv32i/src/I-MISALIGN_LDST-02.S]: Sanity check is OK and I would preserve it. 

My point was, that if this load fails, trap will emulate it and test will pass. In other words, in this implementation there is no check that the load really passed. In original test, if this load fails, test will fail too.

**Comment by allenjbaum** [comment]
I think this is the wrong approach.
The intent of the framework parameters is to select different tests depending on the value of its parameters. 

In this case, unprivileged tests should never run tests that have misaligned accesses unless the "misalign" parameter is true - since unprivileged tests can't rely on privileged resources, and trap handlers inherently touch/use privileged resources

Only a test in the privileged test suite would trap on misaligned load and return the correct value.
So: unpriv test suite, misalign=false: only properly aligned test will be run
      unpriv test suite, misalign= true: add misalign load test
          priv test suite, misalign=false: add misalign load test with trap handler
          priv test suite, misalign= true: nothing more needs to be added

So this should really be two different tests, one of which is in the unpriv test suite, and one of which (will be) in the priv test suite. The test framework should be augmented to
 - only execute the misaligned version without trap handler in the unpriv test suite only if "misalign"=true.
 - only execute the misaligned version with      trap handler in the    priv test suite only if "misalign"=false.

**Comment by allenjbaum** [comment]
I'm fine with discussing this in next weeks meeting.
The whole issue of priv mode tests is looming, and anything that traps must
(eventually) be moved to that test suite (in my opinion).
I welcome other opinions, and if there are storng ones we can add it to the
agenda

**Comment by hajekra** [comment]
@neelgala 
> This pull request is to ensure that until that happens, users do have a working test-suite for various implementations. The previous test was compatible only with implementations which do not support "mis-aligned load/stores" in HW and require to trap. The test I have proposed avoids this limitation.

I agree that we need to have tests working with all possible configurations and implementations.
Personally, I don't think that changing the test (before discussion) to pass all configurations is good way. We can temporary disable this test (and other tests too if needed - misalign jmp and maybe others?) in makefrag.

By the way, in Codasip Framework this is not issue any more, because this particular test is skipped when misalign LDST is implemented in HW.

**Comment by jeremybennett** [comment]
We agreed to commit this last Wednesday, but it has some conflicts that need resolving. Could you deal with these and I can then commit.

Thanks

**Comment by neelgala** [comment]
@jeremybennett 
I have rebased my branch onto the master.
Not sure why github is still showing conflicts. I can't view them since I don't have write-access.
Is it possible for you to resolve the conflicts locally and merge?

The Changelog shows the updates made by this request

**Comment by neelgala** [comment]
Will have to do another pull to rebase with new changes on master. Closing this for now.

**Comment by benjaminselfridge** [comment]
Did this ever get resolved?

**Comment by allenjbaum** [comment]
I just modified the unaligned_jmp test to work regardless of whether unaligned support is there or not.  (Pull reg #54). **I have NOT tested nor even assembled it !!!**

Based on Krste's comments, he wants to ensure that unaligned can be supported in SW, and that requires us to deliberately trap, and have a trap handler which saves the relevant information in the signature, and then returns to the test. The current tests do that, but can't handle the case where there is misalign support, so I added that to the test. 
Now, supported jumps will go to the misaligned address, 
while  the trap handler will return to the *next* aligned address (2 bytes further on). 
At the misaligned address, there is a compressed op that overwrites the signature register, allowing the two cases to be distinguished.

**Pull request closed without merging**